### PR TITLE
Completion field to support multiple completion multi-fields

### LINKED ISSUE #83534: Completion field can't hold more than one completion multi-field
Completion fields support multi-fields. When a multi-field is configured that supports the object format, and not only reading the value, for instance another completion field, an error is thrown when parsing the incoming document.

```
PUT completion_test_index_1
{
  "mappings": {
    "properties": {
      "keyword": {
        "type": "completion",
        "fields": {
          "subfield1": {
            "type": "completion"
          },
          "subfield2": {
            "type": "completion"
          }
        }
      }
    }
  }
}

POST completion_test_index_1/_doc
{
  "keyword": 
    {
      "input": "test,
      "weight": "10"
    }
}
```

here is the error:

```
"root_cause" : [
      {
        "type" : "parsing_exception",
        "reason" : "failed to parse [dummy_field]: expected text or object, but got END_OBJECT",
        "line" : 0,
        "col" : 0
      }
    ]
```
The issue is that we switch the xcontent parser to be able to read the whole object structure of the field multiple times, but we end up exhausting it after the first pass and we stop at its END_OBJECT when trying to read it for the second time as it is not an unexpected token. I suspect configuring a keyword sub-field after the completion sub-field would also cause errors due to the leftover state of the parser.

Another issue that would be nice to address is that we output the error with wrong location and an artificial field name: users may think they have something wrong in their documents but they do not have a `dummy_field` in their documents.

### PR DESCRIPTION
**Pull request by javanna**
The completion field supports multi-fields. In case the sub-field is another completion field, the whole object structure from the incoming document needs to provided as input to the sub-field while parsing it. We have a special XContentParser for this scenario, but it does not handle returning the same object structure multiple times. That is why you can have only one completion sub-field within a completion field, and the error returned when such mechanism breaks is a general parsing error (that mentions a field called dummy_field) that makes users think they have done something wrong in their document.

This commit expands testing for this scenario and extends the parser to support it.

As part of this change, a new parser is created for each sub-field, which makes it possible to expose the same object structure multiple times, for instance in case a completion field has more than one completion sub-fields.

Additionally, the wrapping of both the geo_point multi field parser and the completion multi field parser into a dummy_field object is removed in favour of returning the correct `currentName` of the main field we are parsing.

Additionally getTokenLocation is tweaked to return the location of the field we are parsing in the document, so that error messages become clearer when things go wrong.

Closes #83534

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java [hunk 1]
```diff
@@ -586,28 +589,104 @@ public void doValidate(MappingLookup mappers) {
...
+     * The parser is set to expose by default simple values, unless {@link #nextToken()} is called which is what signals that the
+     * consumer supports the object structure. Note that in order to support multiple completion sub-fields, the inner object parser
+     * needs to be able to return the same object structure multiple times.
+     */
+    /*
+    There are a couple of problems with the current approach: the parser changes behaviour depending on which methods are called by
+    consumers, which is extremely delicate. This kind of works for our internal mappers, but what about mappers from plugins?
+    Also, the sequence of tokens returned by this parser is not valid and things like e.g. copyCurrentStructure blow up.
+    Additionally {@link #getTokenLocation} is incorrect, which is reported together with error messages.
+     */
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java:613]: Passing a Map was what we did before and it caused problems of its own.  I like the idea of creating a new parser for each multifield though, that feels a lot cleaner.

[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java:613]: It should be possible to create a new parser for each multi-field, I am working on a follow-up that should allow to remove some of this complexity. Do you mind if we merge this one in the meantime?

File: libs/x-content/src/main/java/org/elasticsearch/xcontent/support/MapXContentParser.java [hunk 2]
```diff
@@ -28,23 +26,10 @@
  */
 public class MapXContentParser extends AbstractXContentParser {
 
-    private XContentType xContentType;
+    private final XContentType xContentType;
     private TokenIterator iterator;
     private boolean closed;
 
-    public static XContentParser wrapObject(Object sourceMap) throws IOException {
```
Comment (javanna) [inline libs/x-content/src/main/java/org/elasticsearch/xcontent/support/MapXContentParser.java:35]: there's a single user of this left, hence I moved it to a private method local to where it's still needed.

File: server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java:610]: with the recent changes, two out of three problems that I listed are solved. The fact that the parser is delicate stays, as it changes behaviour depending on what consumers do, but I think I can accept that given that it's the only one that does this.

File: server/src/main/java/org/elasticsearch/index/mapper/GeoPointFieldMapper.java [hunk 3]
```diff
@@ -215,7 +215,33 @@ protected void index(DocumentParserContext context, GeoPoint geometry) throws IO
             context.doc().add(new StoredField(fieldType().name(), geometry.toString()));
         }
         // TODO phase out geohash (which is currently used in the CompletionSuggester)
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/GeoPointFieldMapper.java:217]: what's the plan here? From the code, it does not seem like we can phase out geohash given it's what any geopoint sub-field reads. I believe the comment is inaccurate as also keyword and text will get the geohash.

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java [hunk 4]
```diff
@@ -507,16 +507,17 @@ private MultiFields(Map<String, FieldMapper> mappers) {
...
             // of the mappings
             if (mappers.isEmpty()) {
                 return;
             }
-
+            DocumentParserContext context = contextSupplier.get();
             context.path().add(mainField.simpleName());
             for (FieldMapper mapper : mappers.values()) {
                 mapper.parse(context);
+                context = contextSupplier.get();
```
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:520]: not sure, the context is the same after all only the inner parser is switched? Am I missing something?

[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:520]: Also, I looked at making this nicer but it complicated things, I am open to ideas.

[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java]: I changed this. It's not fantastic, but I think better than before. We now pass through the original context as well as the supplier.

**Review by romseygeek — APPROVED** [review]
LGTM, thanks for all the iterations

**Pull request merged**